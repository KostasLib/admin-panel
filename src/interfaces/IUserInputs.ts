export interface IUSerInputs {
  name: string;
  email: string;
  phone: string;
  address: string;
  company: string;
}
