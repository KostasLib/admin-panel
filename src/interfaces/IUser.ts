import { IUSerInputs } from './IUserInputs';

export interface IUser extends IUSerInputs {
  id: string;
  photo?: string;
}
