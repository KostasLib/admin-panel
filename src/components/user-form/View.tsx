import React from 'react';
import { IUSerInputs } from '../../interfaces/IUserInputs';
import { Input } from '../input';
import s from './UserForm.module.scss';

interface IProps {
  inputs: IUSerInputs;
  hasChanged: boolean;
  onChange: (name: keyof IUSerInputs) => (e: any) => void;
  onCancel: (e: React.FormEvent<HTMLFormElement>) => void;
  onSave: (e: React.FormEvent<HTMLFormElement>) => void;
}

/** A stateless function component that handles the presentation for the UserForm. */
export const View: React.FC<IProps> = props => {
  const {
    inputs: { name, email, phone, address, company },
    onChange,
    hasChanged,
    onCancel,
    onSave,
  } = props;

  return (
    <form onSubmit={onSave} onReset={onCancel}>
      <Input
        label="Name"
        type="text"
        name="name"
        placeholder="Enter name"
        title="Enter name"
        value={name}
        required
        maxLength={255}
        onChange={onChange('name')}
      />
      <Input
        label="Email address"
        type="email"
        name="email"
        placeholder="Enter email"
        title="Enter email"
        value={email}
        maxLength={255}
        onChange={onChange('email')}
      />
      <Input
        label="Phone"
        type="tel"
        name="phone"
        placeholder="Enter phone"
        title="Enter phone"
        value={phone}
        maxLength={255}
        pattern="^[^a-zA-Z]*$"
        onChange={onChange('phone')}
      />

      <Input
        label="Address"
        type="text"
        name="address"
        placeholder="Enter address"
        title="Enter address"
        value={address}
        maxLength={255}
        onChange={onChange('address')}
      />
      <Input
        label="Company"
        type="text"
        name="company"
        placeholder="Enter company"
        title="Enter company"
        value={company}
        maxLength={255}
        onChange={onChange('company')}
      />

      <div className={s.buttons}>
        {hasChanged && (
          <input className={s.cancel} type="reset" value="Cancel" title="Cancel" />
        )}
        <input
          className={s.save}
          type="submit"
          value="Save"
          title="Save changes"
          disabled={!hasChanged}
        />
      </div>
    </form>
  );
};
