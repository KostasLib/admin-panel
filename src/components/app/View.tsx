import React from 'react';
import { UserList } from '../user-list';
import { UserForm } from '../user-form';
import { IUSerInputs } from '../../interfaces/IUserInputs';
import { IUser } from '../../interfaces/IUser';
import s from './View.module.scss';

interface IProps {
  users: IUser[];
  selected?: string;
  onSelectUser: (id: string) => () => void;
  onSave: (newData: IUSerInputs) => void;
}

/** A stateless function component that handle's the App's presentation layer. */
export const View: React.FC<IProps> = props => {
  const { onSave, users, selected, onSelectUser } = props;

  return (
    <article className={s.panel}>
      <section className={s['squeezeable-section']}>
        <UserList onSelectUser={onSelectUser} users={users} selected={selected} />
      </section>
      <section className={s['form-section']}>
        <UserForm user={users.find(u => u.id === selected)} onSave={onSave} />
      </section>
    </article>
  );
};
