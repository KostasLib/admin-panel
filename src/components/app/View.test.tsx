import React from 'react';
import { shallow } from 'enzyme';
import { View } from './View';
import { IUser } from '../../interfaces/IUser';

describe("The app's <View/> component", () => {
  const mockUsers: IUser[] = [
    {
      id: '5c093af1c6ee9117a581c7d6',
      photo: 'https://randomuser.me/api/portraits/men/40.jpg',
      name: 'Bates Washington',
      company: 'ZOLAREX',
      email: 'bates.washington@zolarex.io',
      phone: '+1 (915) 447-2207',
      address: '958 Brevoort Place, Ona, Maine, 2433',
    },
    {
      id: '5c093af1aeca1bb00607fb2a',
      photo: 'https://randomuser.me/api/portraits/men/1.jpg',
      name: 'Mollie Oneill',
      company: 'VIAGRAND',
      email: 'mollie.oneill@viagrand.biz',
      phone: '+1 (852) 535-3880',
      address: '120 Cedar Street, Mansfield, Kentucky, 8890',
    },
  ];
  const mockSave = jest.fn();
  const mockSelect = jest.fn();

  it("should render the App's view and pass undedfined to <UserForm />", () => {
    const component = shallow(
      <View users={mockUsers} onSave={mockSave} onSelectUser={mockSelect} />,
    );

    expect(component).toMatchSnapshot();
    expect(component.find('UserForm').prop('user')).toBeUndefined();
  });

  it('should find a user by the selected prop and pass it to <UserForm />', () => {
    const component = shallow(
      <View
        users={mockUsers}
        onSave={mockSave}
        onSelectUser={mockSelect}
        selected={mockUsers[0].id}
      />,
    );
    expect(component).toMatchSnapshot();
    expect(component.find('UserForm').prop('user')).toEqual(mockUsers[0]);
  });
});
