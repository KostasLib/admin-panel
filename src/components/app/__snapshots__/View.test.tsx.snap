// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The app's <View/> component should find a user by the selected prop and pass it to <UserForm /> 1`] = `
<article
  className="panel"
>
  <section
    className="squeezeable-section"
  >
    <UserList
      onSelectUser={[MockFunction]}
      selected="5c093af1c6ee9117a581c7d6"
      users={
        Array [
          Object {
            "address": "958 Brevoort Place, Ona, Maine, 2433",
            "company": "ZOLAREX",
            "email": "bates.washington@zolarex.io",
            "id": "5c093af1c6ee9117a581c7d6",
            "name": "Bates Washington",
            "phone": "+1 (915) 447-2207",
            "photo": "https://randomuser.me/api/portraits/men/40.jpg",
          },
          Object {
            "address": "120 Cedar Street, Mansfield, Kentucky, 8890",
            "company": "VIAGRAND",
            "email": "mollie.oneill@viagrand.biz",
            "id": "5c093af1aeca1bb00607fb2a",
            "name": "Mollie Oneill",
            "phone": "+1 (852) 535-3880",
            "photo": "https://randomuser.me/api/portraits/men/1.jpg",
          },
        ]
      }
    />
  </section>
  <section
    className="form-section"
  >
    <UserForm
      onSave={[MockFunction]}
      user={
        Object {
          "address": "958 Brevoort Place, Ona, Maine, 2433",
          "company": "ZOLAREX",
          "email": "bates.washington@zolarex.io",
          "id": "5c093af1c6ee9117a581c7d6",
          "name": "Bates Washington",
          "phone": "+1 (915) 447-2207",
          "photo": "https://randomuser.me/api/portraits/men/40.jpg",
        }
      }
    />
  </section>
</article>
`;

exports[`The app's <View/> component should render the App's view and pass undedfined to <UserForm /> 1`] = `
<article
  className="panel"
>
  <section
    className="squeezeable-section"
  >
    <UserList
      onSelectUser={[MockFunction]}
      users={
        Array [
          Object {
            "address": "958 Brevoort Place, Ona, Maine, 2433",
            "company": "ZOLAREX",
            "email": "bates.washington@zolarex.io",
            "id": "5c093af1c6ee9117a581c7d6",
            "name": "Bates Washington",
            "phone": "+1 (915) 447-2207",
            "photo": "https://randomuser.me/api/portraits/men/40.jpg",
          },
          Object {
            "address": "120 Cedar Street, Mansfield, Kentucky, 8890",
            "company": "VIAGRAND",
            "email": "mollie.oneill@viagrand.biz",
            "id": "5c093af1aeca1bb00607fb2a",
            "name": "Mollie Oneill",
            "phone": "+1 (852) 535-3880",
            "photo": "https://randomuser.me/api/portraits/men/1.jpg",
          },
        ]
      }
    />
  </section>
  <section
    className="form-section"
  >
    <UserForm
      onSave={[MockFunction]}
    />
  </section>
</article>
`;
